README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# LAGOSUSgis

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![CRAN
status](https://www.r-pkg.org/badges/version/LAGOSUSgis)](https://cran.r-project.org/package=LAGOSUSgis)
[![DOI](https://zenodo.org/badge/106293356.svg)](https://zenodo.org/badge/latestdoi/106293356)

Extra functions to interact with the GIS module of LAGOSUS.

## Features

-   **Easy**: Convenience functions allow for straight-forward
    subsetting.

-   **Fast**: Queries are optimized for speed to avoid loading the
    entirety of massive layers.

-   **Flexible** : Custom queries can be constructed using `SQL`
    statements.

## Installation

``` r
remotes::install_github("cont-limno/LAGOSUSgis")
```

## Usage

``` r
library(LAGOSUSgis)
```

### Download data

Until data is posted publicly, place the `LAGOSUSgis` geodatabase file
in the location returned by `lagosusgis_path()`

### List available GIS layers

``` r
library(sf)
sf::st_layers(LAGOSUSgis::lagosusgis_path())
```

<details closed>
<summary>
<span title="Click to Expand"> layers </span>
</summary>

``` r


|name                                  |driver      | features| fields|
|:-------------------------------------|:-----------|--------:|------:|
|TIGER_Coastline                       |OpenFileGDB |     4245|      3|
|US_Box                                |OpenFileGDB |        1|      2|
|County_Coastline                      |OpenFileGDB |     5756|     19|
|Derived_Land_Borders                  |OpenFileGDB |        7|      3|
|US_Countybased_Clip_Polygon           |OpenFileGDB |        1|      3|
|simple_bailey                         |OpenFileGDB |      165|     15|
|simple_mlra                           |OpenFileGDB |      226|     14|
|simple_wwf                            |OpenFileGDB |       44|     12|
|simple_state                          |OpenFileGDB |       49|     15|
|simple_hu4                            |OpenFileGDB |      202|     13|
|simple_hu8                            |OpenFileGDB |     2105|     13|
|simple_county                         |OpenFileGDB |     3106|     16|
|simple_neon                           |OpenFileGDB |       17|     12|
|LAGOS_US_All_Lakes_1ha_AUSTIN20190521 |OpenFileGDB |   479950|     21|
|NHD_Combined_Regions                  |OpenFileGDB |        1|      2|
|state                                 |OpenFileGDB |       49|     18|
|bailey                                |OpenFileGDB |      165|     19|
|mlra                                  |OpenFileGDB |      226|     18|
|neon                                  |OpenFileGDB |       17|     16|
|wwf                                   |OpenFileGDB |       44|     16|
|hu12                                  |OpenFileGDB |    83108|     16|
|hu8                                   |OpenFileGDB |     2105|     16|
|hu4                                   |OpenFileGDB |      202|     17|
|county                                |OpenFileGDB |     3106|     20|
|omernik3                              |OpenFileGDB |       87|     22|
|epanutr                               |OpenFileGDB |        9|     13|
|nws                                   |OpenFileGDB |    27138|     27|
|ws                                    |OpenFileGDB |   478749|     30|
|LAGOS_limno_linked_merged_20191118    |OpenFileGDB |    81618|     15|
|simple_epanutr                        |OpenFileGDB |        9|     13|
|simple_omernik3                       |OpenFileGDB |       87|     22|
|flatbuff100                           |OpenFileGDB |   635747|      3|
|buff100_unflat                        |OpenFileGDB |   833684|      2|
|simple_hu12                           |OpenFileGDB |    83108|     16|
|flatbuff500                           |OpenFileGDB |  1533137|      3|
|buff500_unflat                        |OpenFileGDB |  4333832|      2|
|Great_Lakes                           |OpenFileGDB |       31|     13|
|simple_buff100                        |OpenFileGDB |   479950|      3|
|simple_buff500                        |OpenFileGDB |   479950|      3|
|LAGOS_US_All_Lakes_1ha_No_Islands     |OpenFileGDB |   479950|      4|
|simple_nws                            |OpenFileGDB |    27138|     11|
|simple_ws                             |OpenFileGDB |   478749|     12|
|buff100                               |OpenFileGDB |   479950|      6|
|buff500                               |OpenFileGDB |   479950|      6|
|lake_as_point                         |OpenFileGDB |   479950|      3|
|LAGOS_US_All_Lakes_1ha_points         |OpenFileGDB |   479950|     71|
|LAGOS_US_All_Lakes_1ha                |OpenFileGDB |   479950|     77|
```

</details>

<br>

### Query a lake polygon

``` r
res_lake <- query_lake_poly(34352)
```

### Query from a specfic layer

``` r
res_iws  <- query_gis(layer = "ws", 
                      id_name = "lagoslakeid", ids = c(34352))
res_lake <- query_gis(layer = "LAGOS_US_All_Lakes_1ha", 
                      id_name = "lagoslakeid", ids = 34352)
res_pnt  <- query_gis(layer = "LAGOS_US_All_Lakes_1ha_POINTS", 
                      id_name = "lagoslakeid", ids = 34352)
```

### Query a combined watershed and lake polygon

``` r
res <- query_wbd(lagoslakeid = c(7010))
```

### Flexible queries using `SQL` statements

``` r
res <- query_gis_(query = "SELECT * FROM ws WHERE lagoslakeid IN ('7010')")
res <- query_gis_(query = paste0("SELECT * FROM LAGOS_US_All_Lakes_1ha_points WHERE lake_centroidstate LIKE '", "RI", "' AND lake_totalarea_ha > 4"))
```

## References

Soranno P., K. Cheruvelil. 2017. LAGOS-NE-GIS v1.0: A module for
LAGOS-NE, a multi-scaled geospatial and temporal database of lake
ecological context and water quality for thousands of U.S. Lakes:
2013-1925. Environmental Data Initiative.
<http://dx.doi.org/10.6073/pasta/fb4f5687339bec467ce0ed1ea0b5f0ca>.
Dataset accessed 9/26/2017.


NEWS.md
# LAGOSUSgis 0.1.x

* Version bump for GIS module v0.9

# LAGOSUSgis 0.1.2

* Convert internal use of the `vapour` package to use `sf` `query` and `wkt_filter`

# LAGOSUSgis 0.1.1

* Version bump for GIS module v0.7

# LAGOSUSgis 0.1.0

* Initial release for GIS module v0.6


tests/testthat.R
library(testthat)
library(LAGOSUSgis)

test_check("LAGOSUSgis")


tests/testthat/test-query.R
test_that("query works", {
  expect_s3_class(
    query_gis("state"),
    "data.frame"
  )
})

test_that("query works with id info", {
  expect_s3_class(
    query_gis("state", "state_name", "Michigan"),
    "data.frame"
  )
})

test_that("query works with extent info", {
  lake_coordinates <- query_gis("LAGOS_US_All_Lakes_1ha_points",
                                "lagoslakeid", "1")

  expect_s3_class(
    query_gis("LAGOS_US_All_Lakes_1ha",
              extent = sf::st_bbox(lake_coordinates)),
    "data.frame"
  )

  expect_s3_class(
    query_gis_(query = "SELECT * FROM LAGOS_US_All_Lakes_1ha",
               extent = sf::st_bbox(lake_coordinates)),
    "data.frame"
  )
})

test_that("query_ works", {
  expect_s3_class(
    query_gis_(query = "SELECT * FROM ws WHERE lagoslakeid IN ('7010')"),
    "data.frame"
  )
})


R/get.R
#' Get LAGOSNEgis data
#'
#' Get data from the GIS module of LAGOSNE
#'
#' @importFrom utils unzip download.file
#' @importFrom gdalUtilities ogr2ogr
#' @param dest_folder file.path defaults to lagosnegis_path()
#' @param overwrite logical
#' @references Soranno P., K. Cheruvelil. 2017. LAGOS-NE-GIS v1.0: A module for
#'  LAGOS-NE, a multi-scaled geospatial and temporal database of lake
#'  ecological context and water quality for thousands of U.S. Lakes:
#'  2013-1925. Environmental Data Initiative.
#'  http://dx.doi.org/10.6073/pasta/fb4f5687339bec467ce0ed1ea0b5f0ca.
#'  Dataset accessed 9/26/2017.
#'
#' @examples \dontrun{
#' lagosnegis_get()
#' }
lagosnegis_get <- function(dest_folder = lagosusgis_dir(), overwrite = FALSE){
  # dest_folder <- LAGOSNEgis:::lagosnegis_path()
  outpath <- file.path(dest_folder, "lagos-ne_gis.gpkg")
  dest_gdb <- file.path(dest_folder, "LAGOS_NE_GIS_Data_v1.0.gdb")

  if(file.exists(outpath) & !overwrite){
    warning("LAGOSNEgis data already exists on the local machine.
  Re-download if neccessary using the 'overwrite` argument.'")
    return(invisible("LAGOS is the best"))
  }

  edi_baseurl   <- "https://portal.edirepository.org/nis/dataviewer?packageid="
  pasta_baseurl <- "http://pasta.lternet.edu/package/data/eml/edi/"
  edi_url       <- paste0(edi_baseurl, c("edi.98.1"))
  pasta_url     <- paste0(pasta_baseurl, "98/1")

  files      <- suppressWarnings(paste0(edi_url, "&entityid=",
                                        readLines(pasta_url)))
  # file_names <- sapply(files, get_file_names)
  # file_ind <- as.numeric(which(!is.na(file_names)))
  files      <- files[20] # files[file_ind]
  file_names <- "LAGOSNE_GIS_Data_v1.0_gdb.zip" # file_names[file_ind]

  local_dir   <- file.path(tempdir())
  dir.create(local_dir, showWarnings = FALSE)
  file_paths  <- file.path(local_dir, file_names)

  invisible(lapply(seq_len(length(files)),
                   function(i) get_if_not_exists(files[i], file_paths[i],
                                                 overwrite)))

  message("LAGOSNEgis downloaded. Now converting to gpkg ...")
  dir.create(dest_folder, showWarnings = FALSE)
  unlink(dest_gdb, recursive = TRUE)
  system(paste0(has_7z()$path, " e ", file_paths, " -o",
                dest_gdb))

  unlink(outpath, recursive = TRUE)
  gdalUtilities::ogr2ogr(dest_gdb, outpath, dim = 2,
                         overwrite = TRUE,
                         nlt = "PROMOTE_TO_MULTI", f = "GPKG")

  gdalUtilities::ogr2ogr(dest_gdb, outpath, dim = 2,
                         sql = "select *, cast(Shape_Area as numeric(30,5)) as Shape_Area from IWS",
                         dialect = "ogrsql", overwrite = TRUE, f = "GPKG")

  # gdalUtils::ogrinfo(lagosnegis_path(), so = TRUE, sql = 'select * from IWS limit 1')
  # gdalUtils::ogrinfo(lagosnegis_path())

  return(invisible(outpath))
}


R/utils.R
albers_conic <- function() {
  # Albers Equal Area Conic
  "+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=23 +lon_0=-96 +x_0=0 +y_0=0 +datum=NAD83 +units=m +no_defs"
}

#' Get the path to LAGOSUSgis data
#'
#' @export
#'
#' @examples \dontrun{
#' # use defaults
#' lagosusgis_path()
#'
#' # custom setting
#' Sys.setenv(
#'   LAGOSUSGIS_PATH = path.expand("~/.local/share/LAGOS-GIS/gis_locus_v1.0.gpkg"))
#' lagosusgis_path()
#'
#' # revert to defaults
#' Sys.setenv(LAGOSUSGIS_PATH = "")
#' lagosusgis_path()
#' }
lagosusgis_path <- function() {
  res_default <- file.path(
    rappdirs::user_data_dir("LAGOS-GIS"), "LAGOS_US_GIS_Data_v0.9.gdb")

  env_vars <- Sys.getenv("LAGOSUSGIS_PATH")

  if (any(!env_vars == "")) {
    res <- env_vars
  } else {
    res <- res_default
  }

  res
}

lagosusgis_dir <- function() {
  rappdirs::user_data_dir("LAGOS-GIS")
}

#' @importFrom curl curl_fetch_memory
#' @importFrom stringr str_extract
get_file_names <- function(url) {
  handle <- curl::new_handle(nobody = TRUE)

  headers <- curl::parse_headers(
    curl::curl_fetch_memory(url, handle)$headers)
  fname <- headers[grep("filename", headers)]

  res <- stringr::str_extract(fname, "(?<=\\=)(.*?)\\_gdb.zip")
  gsub('\\"', "", res)
}

get_if_not_exists <- function(url, destfile, overwrite) {
  if (!file.exists(destfile) | overwrite) {
    download.file(url, destfile, mode = "wb")
  } else {
    message(paste0("A local copy of ", url, " already exists on disk"))
  }
}

#' @importFrom sf st_transform st_crs st_is_longlat st_coordinates st_centroid
toUTM <- function(sf_object) {

  if (is.na(st_crs(sf_object)$epsg)) {
    sf_object <- st_transform(sf_object, crs = 4326)
  }

  if (sf::st_is_longlat(sf_object)) {
    suppressWarnings(
      utm_zone <- long2UTM(st_coordinates(st_centroid(st_union(sf_object)))[1]))
    crs      <- paste0("+proj=utm +zone=", utm_zone, " +datum=WGS84")

    sf::st_transform(sf_object, crs = crs)
  } else {
    sf_object
  }
}

long2UTM <- function(long) {
  (floor((long + 180) / 6) %% 60) + 1
}

has_7z <- function() {
  paths_7z <- c("7z",
    "~/usr/bin/7z",
    "C:\\PROGRA~1\\7-Zip\\7za",
    "C:\\PROGRA~1\\7-Zip\\7z.exe")
  if (!any(nchar(Sys.which(paths_7z)) > 0)) {
    list(yes = FALSE, path = NA)
  } else {
    list(yes = TRUE, path = paths_7z[nchar(Sys.which(paths_7z)) > 0])
  }
}

#' Gets a file from a github repo, using the Data API blob endpoint
#'
#' This avoids the 1MB limit of the content API and uses [gh::gh] to deal with
#' authorization and such.  See https://developer.github.com/v3/git/blobs/ and https://gist.github.com/noamross/73944d85cad545ae89efaa4d90b049db
#' @param url the URL of the file to download via API, of the form
#'   `:owner/:repo/blob/:path
#' @param ref the reference of a commit: a branch name, tag, or commit SHA
#' @param owner,repo,path alternate way to specify the file.  These will
#'   override values in `url`
#' @param to_disk,destfile write file to disk (default=TRUE)?  If so, use the
#'   name in `destfile`, or the original filename by default
#' @param .token,.api_url,.method,.limit,.send_headers arguments passed on to
#'   [gh::gh]
#' @importFrom gh gh
#' @importFrom stringi stri_match_all_regex
#' @importFrom purrr %||% keep
#' @importFrom base64enc  base64decode
#' @export
#' @return Either the local path of the downloaded file (default), or a raw
#'   vector
gh_file <- function(url = NULL, ref = NULL,
                    owner = NULL, repo = NULL, path = NULL,
                    to_disk = TRUE, destfile = NULL,
                    .token = NULL, .api_url = NULL, .method = "GET",
                    .limit = NULL, .send_headers = NULL) {
  if (!is.null(url)) {
    matches <- stringi::stri_match_all_regex(
      url,
      "(github\\.com/)?([^\\/]+)/([^\\/]+)/[^\\/]+/([^\\/]+)/([^\\?]+)"
    )
    owner <- owner %||% matches[[1]][3]
    repo <- repo %||% matches[[1]][4]
    ref <- ref %||% matches[[1]][5]
    path <- path %||% matches[[1]][6]
    pathfile <- basename(path)
  }
  pathdir <- dirname(path)
  if (length(grep("/", path)) == 0) {
    pathdir <- NULL
  }

  dir_contents <- gh(
    "/repos/:owner/:repo/contents/",
    owner = owner, repo = repo
  )

  file_sha <- keep(dir_contents, ~ .$path == path)[[1]]$sha
  blob <- gh(
    "/repos/:owner/:repo/git/blobs/:sha",
    owner = owner, repo = repo, sha = file_sha,
    .token = NULL, .api_url = NULL, .method = "GET",
    .limit = NULL, .send_headers = NULL
  )
  raw <- base64decode(blob[["content"]])
  if (to_disk) {
    destfile <- destfile %||% pathfile
    writeBin(raw, con = destfile)
    return(destfile)
  } else {
    return(raw)
  }
}

R/query.R
#' Query LAGOS GIS
#'
#' @param layer character layer name
#' @param id_name selection column
#' @param ids feature ids to select
#' @param extent apply an arbitrary extent using an sf bbox
#' @param crs character projection info defaults to lagosusgis_path()
#' @param gis_path character path to LAGOSUS GIS gpkg
#'
#' @importFrom sf st_geometry st_geometry_type st_cast st_read st_as_text
#' @importFrom memoise memoise
#' @export
#'
#' @examples \dontrun{
#'
#' sf::st_layers(lagosusgis_path())
#' gdalUtils::ogrinfo(lagosusgis_path(), "hu12", so = TRUE)
#'
#' states   <- query_gis("state")
#'
#' library(mapview)
#' state    <- query_gis("state", "state_name", "Michigan")
#' res_iws  <- query_gis("ws", "lagoslakeid", c(34352))
#' res_lake <- query_gis("LAGOS_US_All_Lakes_1ha", "lagoslakeid", 34352)
#' res_pnt  <- query_gis("LAGOS_US_All_Lakes_1ha_points", "lagoslakeid", 34352)
#' mapview(state) + mapview(res_iws) + mapview(res_lake) + mapview(res_pnt)
#'
#' res <- query_gis("ws", "lagoslakeid", c(7010))
#' res <- query_gis("hu12", "hu12_zoneid", c("hu12_1"))
#' res <- query_gis("hu8", "hu8_zoneid", c("hu8_100"))
#' res <- query_gis("hu4", "hu4_zoneid", c("hu4_5"))
#'
#' # query multiple feature ids
#' res <- query_gis("ws", "lagoslakeid", c(7010, 1))
#'
#' # query lake polygon from coordinates using extent argument
#' lake_coordinates <- query_gis("LAGOS_US_All_Lakes_1ha_points",
#'   "lagoslakeid", "1")
#'
#' lake_polygon <- query_gis("LAGOS_US_All_Lakes_1ha",
#'   extent = sf::st_bbox(lake_coordinates))
#' }
query_gis <- function(layer, id_name = NULL, ids = NULL, extent = character(0),
                      crs = albers_conic(),
                      gis_path = lagosusgis_path()) {

  if (all(is.null(id_name), is.null(ids))) {
    if (length(extent) > 0) { # extract by extent (bbox)
      wkt_filter <- sf::st_as_sfc(extent, crs = albers_conic())
      res <- sf::st_read(gis_path, layer = layer,
        wkt_filter = sf::st_as_text(wkt_filter), quiet = TRUE)
    } else { # load entire layer
      res <- sf::st_read(gis_path, layer = layer, quiet = TRUE)
    }
  } else { # load specific ids from specific layer
    res <- query_gis_(gis_path,
      query = paste0("SELECT * FROM ", layer,
        " WHERE ", id_name, " IN ('",
        paste0(ids,
          collapse = "', '"), "')"),
      extent = extent,
      crs = crs)

    # sort items by ids
    res <- res[match(ids, data.frame(res[, id_name])[, id_name]), ]
  }

  if (any(unique(sf::st_geometry_type(sf::st_geometry(res))) == "MULTISURFACE")) {
    res <- sf::st_cast(res, "MULTIPOLYGON")
  }

  res
}

#' Raw (non-vectorized) query of LAGOS GIS
#'
#' @param gis_path file path
#' @param query SQL string
#' @param extent apply an arbitrary extent using an sf bbox
#' @param crs coordinate reference system string or epsg code
#'
#' @importFrom dplyr mutate select
#' @importFrom sf st_as_sfc st_crs st_geometry st_zm
#' @importFrom rlang .data
#'
#' @export
#'
#' @examples \dontrun{
#' res <- query_gis_(query = "SELECT * FROM ws WHERE lagoslakeid IN ('7010')")
#'
#' # query nested hucs
#' hu4  <- query_gis("hu4", "hu4_huc4", c("0415"))
#' hu8s <- query_gis_(query = "SELECT * FROM hu8 WHERE hu8_huc8 LIKE '0415%'")
#'
#' # query multiple nested hucs
#' hu4s  <- query_gis("hu4", "hu4_huc4", c("0415", "0414"))
#' hu8s  <- query_gis_(query = paste0("SELECT * FROM hu8 WHERE ",
#'   paste0("hu8_huc8 LIKE '", hu4s$hu4_huc4, "%'", collapse = " OR ")))
#'
#' # query lake polygon from coordinates using extent argument
#' lake_coordinates <- query_gis("LAGOS_US_All_Lakes_1ha_points",
#'   "lagoslakeid", "1")
#' lake_polygon <- query_gis_(query = "SELECT * FROM LAGOS_US_All_Lakes_1ha",
#'   extent = sf::st_bbox(lake_coordinates))
#' }
#'
query_gis_ <- function(gis_path = lagosusgis_path(), query, extent = character(0),
                       crs = albers_conic()) {

  # error if extent is not NA and query is defined?

  # investigate specific layers
  # library(sf)
  # crs <- LAGOSUSgis:::albers_conic()
  # gis_path <- path.expand("~/.local/share/LAGOS-GIS/LAGOS_US_GIS_Data_v0.7.gdb")
  # sf::st_layers(gis_path)

  if (!file.exists(gis_path) & !dir.exists(gis_path)) {
    stop(paste0("Data not available at the path pointed to by lagosusgis_path():\n  '",
      lagosusgis_path(), "'"))
  }

  ###
  layer <- stringr::str_extract(query, "(?<=FROM )(.*)(?= WHERE)")
  dat <- sf::st_read(gis_path, query = query,
    wkt_filter = sf::st_as_text(sf::st_as_sfc(extent)),
    quiet = TRUE)

  # sf::st_crs(dat)      <- sf::st_crs(crs)
  #
  # if(any(unique(sf::st_geometry_type(sf::st_geometry(dat))) == "MULTISURFACE")){
  #   dat <- sf::st_cast(dat, "MULTIPOLYGON")
  # }

  sf::st_zm(dat)
}

#' Query watershed boundary for LAGOS lakes
#'
#' @param lagoslakeid numeric
#' @param gis_path file.path to LAGOSUS GIS gpkg
#' @param crs projection string or epsg code
#' @param utm logical convert crs to utm
#'
#' @importFrom sf st_union st_geometry<- st_crs<- st_buffer
#' @importFrom memoise memoise
#' @export
#' @examples \dontrun{
#' library(mapview)
#' res <- query_wbd(lagoslakeid = c(7010))
#' res <- query_wbd(lagoslakeid = c(7010, 34352))
#' res <- query_wbd(lagoslakeid = c(34352))
#' mapview(res)
#'
#' res <- query_wbd(lagoslakeid = c(2057, 3866, 1500, 3386, 2226,
#'   1637, 6874, 7032, 1935, 6970, 5331, 34352))
#' res <- res[res$lagoslakeid == 34352, ]
#' mapview::mapview(res)
#' }
query_wbd <- function(lagoslakeid, gis_path = lagosusgis_path(),
                      crs = albers_conic(), utm = FALSE) {

  iws           <- query_gis("ws", "lagoslakeid", lagoslakeid)
  lake_boundary <- query_gis("LAGOS_US_All_Lakes_1ha", "lagoslakeid", lagoslakeid)

  res <- lapply(seq_len(nrow(iws)), function(x) {
    iws_dissolve  <- st_buffer(do.call(c, st_geometry(iws)[x]), 0)
    lake_dissolve <- st_buffer(do.call(c, st_geometry(lake_boundary)[x]), 0.001)

    st_geometry(st_union(iws_dissolve, lake_dissolve))
  })

  res <- do.call(c, res)
  st_crs(res) <- crs
  st_geometry(iws) <- res

  if (utm) {
    toUTM(iws)
  } else {
    iws
  }
}

#' Query polygon outline for LAGOS lakes
#'
#' @param lagoslakeid numeric
#' @param gis_path file.path to LAGOSUS GIS gpkg
#' @param crs projection string or epsg code
#' @param utm logical convert crs to utm
#'
#' @export
#'
#' @examples \dontrun{
#' res <- query_lake_poly(c(474805, 474803))
#' }
query_lake_poly <- function(lagoslakeid, gis_path = lagosusgis_path(),
                            crs = albers_conic(), utm = FALSE) {

  res <- query_gis("LAGOS_US_All_Lakes_1ha", "lagoslakeid", lagoslakeid)

  if (utm) {
    toUTM(res)
  } else {
    res
  }
}